### PR TITLE
Avoid in-place ops during logging result updates

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/11386

This is caused by:

```python
import torch

a = torch.tensor(0.0, dtype=torch.float)
b = torch.tensor(3.0, dtype=torch.double)

a = a + b  # same as `a.add(b)`
# `a` promoted to double
assert a.dtype is torch.double

a = torch.tensor(0.0, dtype=torch.float)  # reset dtype

a += b  # same as `a.add_(b)`
# `a` NOT promoted to double
assert a.dtype is torch.float
```

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @tchaton @rohitgr7 @carmocca @edward-io @ananthsub @kamil-kaczmarek @Raalsky @Blaizzy

### CODE REVIEW

**Review by rohitgr7 — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !
